### PR TITLE
extricate location.d into its own file

### PR DESCRIPTION
**Pull request by WalterBright**
Pulling out `struct Loc`. This will help in disentangling modules that need Loc but nothing else from globals.d.

Cut&paste job for easy review and first stab at this.

### CODE REVIEW

**Comment by rikkimax** [comment]
https://github.com/dlang/dmd/blob/master/dub.sdl#L33

Will need it added here (unfortunately that the dub file has to be file-based instead of just using directories since it isn't one directory per sub package).

File: compiler/src/dmd/location.d [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        import core.stdc.string : strcmp;
+
+        return charnum == loc.charnum &&
+               linnum == loc.linnum &&
+               (filename == loc.filename ||
+                (filename && loc.filename && strcmp(filename, loc.filename) == 0));
+    }
+
+    /// ditto
+    extern (D) size_t toHash() const @trusted pure nothrow
```
Comment (rikkimax) [inline compiler/src/dmd/location.d:118]: oooo, this could be @nogc!

File: compiler/src/dmd/location.d [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    /// zero-terminated filename string, either absolute or relative to cwd
+    const(char)* filename;
+    uint linnum; /// line number, starting from 1
+    uint charnum; /// utf8 code unit index relative to start of line, starting from 1
+    version (LocOffset)
+        uint fileOffset; /// utf8 code unit index relative to start of file, starting from 0
+
+    static immutable Loc initial; /// use for default initialization of const ref Loc's
+
+nothrow:
```
Comment (rikkimax) [inline compiler/src/dmd/location.d:40]: Lots of methods have nothrow on them, as well as this at global scope, there could be some simplification here!

**Comment by WalterBright** [comment]
```
... runnable\b26.d                 ()
Failed to execute the `unit` test executable! (exit code -1073741819)

> D:\a\1\s\generated\windows\release\64\dmd.exe -version=NoBackend -version=GC -version=NoMain -version=MARS -version=DMDLIB -unittest -JD:\a\1\s\generated\windows\release\64 -Jsrc/dmd/res -Isrc -ID:\a\1\s\compiler\test\unit -i -main -ofD:\a\1\s\compiler\test\test_results\runner.exe -m64 test\unit\compilable\crlf.d test\unit\compilable\searching.d test\unit\deinitialization.d test\unit\frontend.d test\unit\interfaces\check_implementations_20861.d test\unit\lexer\diagnostic_reporter.d test\unit\lexer\lexer_dmdlib.d test\unit\lexer\location_offset.d test\unit\objc\protocols\diagnostic_messages.d test\unit\objc\protocols\optional_methods.d test\unit\parser\aliasdeclaration_location.d test\unit\parser\conditionalcompilation_location.d test\unit\parser\diagnostic_reporter.d test\unit\parser\dvcondition_location.d test\unit\parser\linkage_location.d test\unit\parser\visibilitydeclaration_location.d test\unit\self_test.d test\unit\semantic\control_flow.d test\unit\semantic\covariance.d test\unit\semantic\imports.d test\unit\support.d test\unit\triple.d D:\a\1\s\compiler\test\test_results\runner.d -g
D:\a\1\s\compiler\test\..\..\..\phobos\std\traits.d(8514): Deprecation: `__traits(getAttributes)` may only be used for individual functions, not overload sets such as: `executeTest`

D:\a\1\s\compiler\test\..\..\..\phobos\std\traits.d(8514):        the result of `__traits(getOverloads)` may be used to select the desired function to extract attributes from

D:\a\1\s\compiler\test\..\..\..\phobos\std\traits.d(8514): Deprecation: `__traits(getAttributes)` may only be used for individual functions, not overload sets such as: `executeTest`

D:\a\1\s\compiler\test\..\..\..\phobos\std\traits.d(8514):        the result of `__traits(getOverloads)` may be used to select the desired function to extract attributes from

D:\a\1\s\compiler\test\..\..\..\phobos\std\traits.d(8514): Deprecation: `__traits(getAttributes)` may only be used for individual functions, not overload sets such as: `toString`

D:\a\1\s\compiler\test\..\..\..\phobos\std\traits.d(8514):        the result of `__traits(getOverloads)` may be used to select the desired function to extract attributes from

D:\a\1\s\compiler\test\..\..\..\phobos\std\traits.d(8514): Deprecation: `__traits(getAttributes)` may only be used for individual functions, not overload sets such as: `toString`

D:\a\1\s\compiler\test\..\..\..\phobos\std\traits.d(8514):        the result of `__traits(getOverloads)` may be used to select the desired function to extract attributes from


> D:\a\1\s\compiler\test\test_results\runner.exe
test.d(1): Error: declaration expected following attribute, not end of file

```
```
https://dev.azure.com/dlanguage/dmd/_build/results?buildId=34370&view=logs&j=6775e4b2-0d76-5db7-9bcd-015c7821c824&t=1bf40372-f529-586c-b834-12b53d9adf74&l=830
```
I have no idea whatsoever what this error is in the Azure pipeline tests or what file `test.d` is or what it might have to do with this PR.

**Comment by WalterBright** [comment]
I guess I found that problem! Anyhow, all those `import globals : Loc;` statements justify this PR.

File: compiler/src/dmd/location.d (hunk 2 above)
Comment (WalterBright) [reply compiler/src/dmd/location.d:40]: I know, but pls save it for a later PR.

**Comment by WalterBright** [comment]
Ready to go

File: compiler/src/dmd/README.md [hunk 3]
```diff
@@ -47,6 +47,7 @@ Note that these groups have no strict meaning, the category assignments are a bi
 | File                                                                  | Purpose                                                              |
 |-----------------------------------------------------------------------|----------------------------------------------------------------------|
 | [lexer.d](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/lexer.d)   | Convert source code into tokens for the D and ImportC parsers        |
+| [lexer.d](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/location.d)| Encapsulate file/line/column info for error messages, etc.           |
```
Comment (thewilsonator) [inline compiler/src/dmd/README.md:50]: wrong text `[lexer.d]` should be `[location.d]`

**Review by thewilsonator — DISMISSED** [review]
I'm not convinced this is a good move, for it to be worth doing I would have expected many more removals of `import dmd.globals;` and this add _yet even more files_ without packaging modules properly. 

Unless this in pursuance of that IMO this should be left as is.

File: compiler/src/dmd/README.md (hunk 3 above)
Comment (WalterBright) [reply compiler/src/dmd/README.md]: ok

**Comment by WalterBright** [comment]
I noticed several more places where `global.d` was imported, but only `Loc` was used. I did not add them to this PR in order to make this PR less complicated. This is in addition to the dozen or so removals of `import globals : Loc" in this PR.

This is to remove the dependency of identifier.d, tokens.d, lexer.d, parse.d and cparse.d on globals.d, which will make things more modular and cut down on the thicket of interdependencies. The short term goal is to make it possible to use the lexer without having to link in the rest of D. The next step is to do that for the parser. This is a necessary step to making it easier to build a compiler-as-service which is needed for modern IDE integration.

**Comment by ibuclaw** [comment]
Will we finally make Loc a uint? :-)

**Comment by WalterBright** [comment]
There's always somebody with 65,000 files and some with 65,000 lines in them.

**Comment by ibuclaw** [comment]
> There's always somebody with 65,000 files and some with 65,000 lines in them.

ulong is also much better than the current situation too, you know.

File: compiler/src/dmd/location.d [hunk 4]
```diff
@@ -0,0 +1,136 @@
+/**
+ * Encapsulates file/line/column locations.
+ *
+ * Copyright:   Copyright (C) 1999-2022 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 https://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 https://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/globals.d, _location.d)
```
Comment (Geod24) [inline compiler/src/dmd/location.d:7]: Wrong link

File: compiler/src/dmd/location.d [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+ * Documentation:  https://dlang.org/phobos/dmd_location.html
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/location.d
+ */
+
+module dmd.location;
+
+import dmd.common.outbuffer;
+import dmd.root.filename;
+import dmd.globals;
```
Comment (Geod24) [inline compiler/src/dmd/location.d:16]: Why the circular import ?

**Review by Geod24 — APPROVED** [review]
Neat, just one issue and one typo, but glad we are doing this.

**Comment by maxhaton** [comment]
Should `location` go into root at some point? The backend also needs locations for debug info

File: compiler/src/dmd/location.d (hunk 4 above)
Comment (WalterBright) [reply compiler/src/dmd/location.d]: fixed

File: compiler/src/dmd/location.d (hunk 5 above)
Comment (WalterBright) [reply compiler/src/dmd/location.d:16]: Because I haven't yet completely disentangled location from globals, to keep the PR simple.

**Comment by WalterBright** [comment]
@maxhaton a topic for another PR

**Pull request merged**